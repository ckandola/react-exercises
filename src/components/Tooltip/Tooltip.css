:root {
    --transition: all 0.3s linear;
}

.tooltip-main {
    display:none;
}

.tooltip-main-show {
    position:relative;
    width:fit-content;
    color:white;
    display:block;
    z-index:10;
    padding:1em;
    border-radius: 1em;
    background: black;
    transition:var(--transition);
}

.tooltip_top::after {
    content: '';
    border-left: 5px solid transparent;
    border-right: 5px solid transparent;
    border-top: 5px solid black;
    position: absolute;
    top: 3.5em;
    left:10%;
}

.tooltip_bottom::before {
    content: '';
    display: block;
    border-left: 5px solid transparent;
    border-right: 5px solid transparent;
    border-bottom: 5px solid black;
    position: absolute;
    transform: translateX(-50%);
    top: -0.3em;
    left:10%;
}

