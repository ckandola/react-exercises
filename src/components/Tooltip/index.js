import React, {useState, useRef} from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';

const Tooltip = ({text, canToggleWithClick = true, textPosition = 'bottom', children}) => {
    const [opacity, setOpacity]= useState(false);
    const [position, setPosition] = useState({top:1, left:1});
    const nodeRef = useRef();

    const toggle = isMouseOut => {
        const tooltipNode = nodeRef.current;
        
        setPosition({top: tooltipNode.offSetTop, left: tooltipNode.left})
        if(canToggleWithClick){
            // Only opens tooltip if text has been clicked, ensures that it is closed when not hovering
            if(!isMouseOut || (isMouseOut && opacity)){
                setOpacity(!opacity);
            }
        }
        else {
            // if we can't toggle with a click, just alternate opacity when hovering and when not
            setOpacity(!opacity);
        }
    }

    const handleClick = () => {
        if(canToggleWithClick) {
            toggle(false);
        }
    }

    const top = position.top || 0;
    const style = {
        zIndex: opacity ? 1000: -1000,
        top: top + (textPosition === 'bottom' ? + 20 : - 60),
        left: (position.left || 0),
        opacity: +opacity,
    };
    const tooltipClasses = 'tooltip ' + textPosition;

    return (
        <div ref={nodeRef} 
            style={{display: 'inline'}}>
            <span 
                style={{color: 'blue'}}
                onClick={event => handleClick()}
                onMouseOver={() => toggle(false)}
                onMouseOut={() => {toggle(true)}}
            >{children}</span>
            <div className={tooltipClasses}
                style={style}
                role="tooltip">
                    <div className="tooltip-arrow">
                        <div className="tooltip-inner">
                            {text}
                        </div>
                    </div>
            </div>
        </div>
    )

}

export default Tooltip;