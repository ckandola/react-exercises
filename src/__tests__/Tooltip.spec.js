import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tooltip from '../components/Tooltip/';

describe('Tooltip tests', () => {
    it('matches the snapshot', () => {
        const wrapper = shallow(<Tooltip text={"hi"}>Read this</Tooltip>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('displays text below children when mouse hovers over children', () => {
        const wrapper = shallow(<Tooltip text={"More info!!"}>Hover for more info</Tooltip>)
        expect(wrapper.find('.tooltip-main-show')).toHaveLength(0);
        wrapper.find('span').simulate('mouseover');
        expect(wrapper.find('.tooltip-main-show')).toHaveLength(1);
    });
});
