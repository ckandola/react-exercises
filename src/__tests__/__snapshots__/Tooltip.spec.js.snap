// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip tests matches the snapshot 1`] = `
<div>
  <span
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    Read this
  </span>
  <div>
    <div
      className="tooltip-main"
    >
      hi
    </div>
  </div>
</div>
`;
